### ISSUE TITLE
Tied parameters of a model cause error with grad_sampler

### ISSUE THREAD

**Issue by huseyinatahaninan** [pytorch/opacus#413]
## 🐛 Bug

When a model has tied parameters (such as tied embedding layer of GPT-2 model), the current version of Opacus stacks the grad_samples in a list instead of summing them up the gradients coming from shared parameters:
https://github.com/pytorch/opacus/blob/3ca9a67a1784cdb053623f95df8e835cdf3b6d08/opacus/grad_sample/grad_sample_module.py#L53

This later causes an error at: `per_sample_norms = torch.stack(per_param_norms, dim=1).norm(2, dim=1)`
because here all parameters have grad_samples as tensors whereas the tied embedding has it as a list of tensors instead.

In the earlier version of Opacus (v0.15) there was this function `create_or_accumulate_grad_sample` that enabled us to register tied embedding layer for instance as it adds up the gradients:
https://github.com/pytorch/opacus/blob/13cf52de6e967ef9cb1167ad4d4981ee15b45697/opacus/grad_sample/utils.py#L75

When I change the lines 50-53 below in the new version and simply replace them with `p.grad_sample += p._current_grad_sample` it works fine because it adds up the gradients coming from shared parameters.
https://github.com/pytorch/opacus/blob/3ca9a67a1784cdb053623f95df8e835cdf3b6d08/opacus/grad_sample/grad_sample_module.py#L53

I am curious why this change to stacking to list was made and whether I am missing anything regarding the tied parameters case. Is there a better fix I can make without changing the internal code? Any help would be greatly appreciated!

## Please reproduce using our [template Colab](https://colab.research.google.com/drive/1R-dnKipK9LOVV4_oKbuHoq4VvGKGbDnd?usp=sharing) and post here the link
https://colab.research.google.com/drive/1OfD4Ubb010Fe0M4rg13LUytNKPpk_asu?usp=sharing

## To Reproduce

Simply following the colab should show the error:

1. Take GPT-2 model from Huggingface
2. Add DP components
3. Run an arbitrary sample forward-backward and call optimizer
4. Tied parameters have list of grad_samples instead of a single tensor accumulating gradients and this causes the error

## Expected behavior

Previously we were able to register such layers by using `create_or_accumulate_grad_sample` as in v0.15 https://github.com/pytorch/opacus/blob/13cf52de6e967ef9cb1167ad4d4981ee15b45697/opacus/grad_sample/utils.py#L75 which added the grad_samples but now I am unable to find a similar solution.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:

```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by gugarosa** [comment]
If I am not mistaken, such behavior is used to provide support for gradient accumulation. For example, if you were to set the `gradient_accumulation_steps` argument from `huggingface/transformers` to a value higher than `1`, it would not work without the mentioned changes.

Nevertheless, there is a naïve way to make it work without changing any core functions. But please note that it will only work without gradient accumulation.

Just rewrite the `compute_embedding_grad_sample` from `opacus/grad_sample/embedding.py` by adding the following before the `return`:

```
if getattr(layer.weight, 'grad_sample') is not None:
        layer.weight.grad_sample = None
```

This will block tied `nn.Embedding` layers from having a list of `grad_sample`, but it will also block the ability of using gradient accumulation, since we will never enter in the `if p.grad_sample is not None` conditional of function `promote_current_grad_sample`.

**Comment by huseyinatahaninan** [comment]
Thanks very much for your comment, really appreciate it!

I don't think it may be about gradient accumulation actually. Because in each gradient accumulation step we call
`optimizer.signal_skip_step(do_skip=True)`
`optimizer.step()`
`optimizer.zero_grad()`

Note that `optimizer.step()` will call `self.pre_step()` and that will call `self.clip_and_accumulate()`. This function then clips the current `grad_samples` and then accumulates them in another parameter `p.summed_grad += grad`. After this step `optimizer.zero_grad()` clears out the `grad_sample` by having `p.grad_sample = None`. So in any case whether we apply gradient accumulation or not, `grad_sample`s are processed and cleared in steps between.

So in my opinion, when there is tied parameters, stacking them into a list in a single forward-backward pass is not relevant to gradient accumulation because at each step of gradient accumulation, `grad_samples` are already processed and cleared out. So I am not really sure why stacking them into a list is done. If I am not missing anything, my fix would also be working fine with gradient accumulation in my opinion.

I am not sure your suggestion would work for me because in case there are tied parameters, `getattr(layer.weight, 'grad_sample')` will return True and then we'd be wiping off the gradients coming from shared parameters. I think that would not work well for the shared parameters case if I understand things correctly.

**Comment by ffuuugor** [comment]
Huh, that's a very interesting use case for sure, thanks for flagging!
Opacus currently doesn't support tied parameters, however I think there's quite a minimal change that would allow it. If it works, I'll try to put up a PR later today. Unfortunately, I don't think any client-side change would help, but fwiw this should work out of the box, so the core fix is due.

The story behind the change, as @gugarosa correctly pointed out, is gradient accumulation. `GradSampleModule` aims to mimic the interface and the behaviour of `nn.Module` where possible. When you do multiple forward+backward passes without calling `zero_grad()` (and `optimizer.step` for that matter), standard pytorch module will sum up the gradients in `p.grad` as if it was one batch. We mimic this behaviour by storing *per sample* gradients in an array - one tensor per physical batch (we could've stored it in a single tensor, but in an array you have a nice introspection which sample belonged to a which batch)
@huseyinatahaninan , the example of accumulation you're describing involve calling the optimizer after every pass (albeit with `do_skip=True`), which is not true for all potential cases. It does make the most sense for the most common situations though, as you're able to save memory while not doing the actual optimization steps, but there are other situations where you wouldn't be calling an optimizer. 

We do, however, have a mechanism to deal with something very similar to tied parameters - recurrent nets. The idea is simple - if gradients belong to the same sample, they need to be summed up. If they belong to different samples, they need to be stacked. For recurrent nets we distinguish between these two situations by looking at the list of activations (`grad_sample_module.py:346`). I think if we slightly modify this condition, it'll also work nicely on parameter level

**Comment by huseyinatahaninan** [comment]
@ffuuugor, thanks very much for your comment! So basically when we apply gradient accumulation, in any case `clip_and_accumulate` should be called because we need to make sure that per-sample gradients are clipped and accumulated in each forward-backward operation. And since it already keeps the accumulated gradients in this other tensor `p.summed_grad`, it already accomplishes mimicking this behavior as in `p.grad` no?

Because of this, I was safely clearing out all `grad_sample`s with `zero_grad()` even during the gradient accumulation. So you are saying while there is nothing wrong with that, there could be cases where you'd not want to clear them out and stack them in a list instead for some reason. Did I understand this correctly?

Your suggestion in the final paragraph sounds like a really neat solution to include tied parameters :) Thanks so much in general for such a great tool for DP training!

**Comment by lxuechen** [comment]
I've been monitoring this thread for quite a while. Just chiming in case my experience could be helpful for you, thus sharing. 

For context, I've been using Opacus on and off since early 2021, and parameter sharing was one of the big issues that took much of my time away. I eventually forked and rewrote part of the lib for this reason (and many others), as much of the grad computation wasn't handled in a way optimized for my problem (language model fine-tuning). 

I've done quite some experiments while writing [this paper](https://openreview.net/pdf?id=bVuP3ltATMz), and it turns that for most fine-tuning problems I tried, there isn't a huge different between tied and untied DP training, if appropriate hyperparameters are applied. 

Note, in the untied case, you'd likely need to manually copy the embedding weights in the lm-head, just to ensure it's not randomly initialized. 

As a bonus point, by untying weights, you get to use our *ghost clipping* algorithm (Section 4 of paper), that saves you a ton of memory during training!

**Comment by ffuuugor** [comment]
> And since it already keeps the accumulated gradients in this other tensor p.summed_grad, it already accomplishes mimicking this behavior as in p.grad no

Not exactly. In the scenario I was describing you don't call `clip_and_accumulate` - this is mostly to support the usage of `GradSampleModule` independently of `DPOptimizer`. Let me give an example scenarios:

## Scenario 1: Stacking accumulation without clipping/aggregation
```
y = model(x0)
loss = criterion(y, labels)
loss.backward()

y = model(x1)
loss = criterion(y, labels)
loss.backward()

optimizer.step()
optimizer.zero_grad()
```

## Scenario 2: Accumulation with clipping
```
y = model(x0)
loss = criterion(y, labels)
loss.backward()

optimizer.signal_skip_step(do_skip=True)
optimizer.step()

y = model(x1)
loss = criterion(y, labels)
loss.backward()

optimizer.step()
optimizer.zero_grad()
```

Both scenarios are valid, but there are differences
1) Scenario 1 will consume more memory, as you're keeping per sample gradient from both x0 and x1 at the same time
2) `expected_batch_size` you should pass to the optimizer is different. In scenario 1, x0 and x1 are considered independent batches and therefore `expected_batch_size` should be the size of either one. Scenario 2 was designed mostly for `BatchMemoryManager` and the definition of the logical batch is the combination of x0 and x1, therefore `expected_batch_size` should be the total size
3) accounting sample rate is not affected, as `step_hook` is not called in between the steps in both scenarios.

> Because of this, I was safely clearing out all grad_samples with zero_grad() even during the gradient accumulation.

There are two things to be aware of
1) `optimizer.zero_grad()` will clear both `grad_sample` and `summed_grad`, unless you've called  `signal_skip_step`, in which case it'll keep `summed_grad`
2) `model.zero_grad()` will always clear only `grad_sample` field, but probably shouldn't be used in a typical training loop, because you want to clear `summed_grad` too after the accumulation is over.

> So you are saying while there is nothing wrong with that, there could be cases where you'd not want to clear them out and stack them in a list instead for some reason. Did I understand this correctly?

yes, that's exactly right

> Your suggestion in the final paragraph sounds like a really neat solution to include tied parameters :) Thanks so much in general for such a great tool for DP training!

Thanks, really appreciate it :)
Feel free to check if the solution in #417 works for you - I think we'll merge it into main soon, but it could take another week or two for it to make its way to the PyPi release

**Comment by huseyinatahaninan** [comment]
Thanks so much for such detailed and clean comment @ffuuugor! Indeed I have always followed Scenario 2 and as you mentioned I did call `signal_skip_step` to make sure only `grad_sample` is cleaned by `optimizer.zero_grad()` and `summed_grad` is kept during gradient accumulation. Glad to hear this is one of the right approaches at least :)

I never thought Scenario 1 would be applied for some reason so I guess that was my confusion about the whole thing. Obviously it could be handy in certain cases though. I understand, thanks very much!

I'll definitely try your suggestion, really appreciate you taking the time to help mitigate this!

@lxuechen thanks very much for chiming in really! I am very well aware of your great work and it's been in my to-do list for a while to checkout your repo and try experimenting with ghost clipping approach as I am also mostly interested in language model fine-tuning :) Thanks very much for the tips you mentioned in your comment :)
